src/wallet/walletServiceStorageProxy.ts

```diff
@@ -0,0 +1,186 @@
+ */
+export class WalletServiceStorageProxy {
+  private wallet: HathorWalletServiceWallet;
+
+  private originalStorage: IStorage;
+
+  constructor(wallet: HathorWalletServiceWallet, originalStorage: IStorage) {
+    this.wallet = wallet;
+    this.originalStorage = originalStorage;
+  }
+
+  /**
+   * Creates a proxy that wraps the original storage with additional methods
+   * needed for nano contract transaction signing.
+   */
+  createProxy(): IStorage {
+    return new Proxy(this.originalStorage, {
+      get: this.proxyHandler.bind(this),
+    });
+  }
+
+  private proxyHandler(target: IStorage, prop: string | symbol, receiver: IStorage): unknown {
+    if (prop === 'getAddressInfo') {
+      return this.getAddressInfo.bind(this);
+    }
+
+    if (prop === 'getTxSignatures') {
+      return this.getTxSignatures.bind(this, receiver);
+    }
+
+    if (prop === 'getTx') {
+      return this.getTx.bind(this);
+    }
+
+    if (prop === 'getSpentTxs') {
+      return this.getSpentTxs.bind(this);
+    }
+
+    if (prop === 'getCurrentAddress') {
+      return this.getCurrentAddress.bind(this);
+    }
+
+    // For all other properties, use the original behavior
+    const value = Reflect.get(target, prop, receiver);
+
+    // Bind methods to maintain correct 'this' context
+    if (typeof value === 'function') {
+      return value.bind(target);
+    }
+
+    return value;
+  }
+
+  /**
+   * Get address information including BIP32 index
+   */
+  private async getAddressInfo(
+    address: string
+  ): Promise<(IAddressInfo & IAddressMetadata & { seqnum: number }) | null> {
+    const addressDetails = await this.wallet.getAddressDetails(address);
+
+    if (!addressDetails) {
+      return null;
+    }
+
+    return {
+      bip32AddressIndex: addressDetails.index,
+      base58: addressDetails.address,
+      seqnum: addressDetails.seqnum,
+      numTransactions: addressDetails.transactions,
+      // TODO: This balance is not used by any of the nano contract methods
+      // but in order to be 100% compatible with the old facade method, we need
+      // to implement an API to fetch the balance for all tokens given an address
+      balance: new Map(),
+    };
+  }
+
+  /**
+   * Get transaction signatures using the transaction utility
+   */
+  // eslint-disable-next-line class-methods-use-this
+  private async getTxSignatures(receiver: IStorage, tx: Transaction, pinCode: string) {
+    const result = await transactionUtils.getSignatureForTx(tx, receiver, pinCode);
+    return result;
+  }
+
+  /**
+   * Get spent transactions for input signing
+   * This is an async generator that yields transaction data for each input
+   */
+  private async *getSpentTxs(inputs: any[]) {
+    for (let index = 0; index < inputs.length; index++) {
+      const input = inputs[index];
+      const tx = await this.getTx(input.hash);
+      if (tx) {
+        yield { tx, input, index };
+      }
+    }
+  }
```

Comment:
Not sure if it's worth implementing bug we could have more than one input of the same tx, so we could have a cache of the requested txs, so we don't request the same one twice.